### PR TITLE
add crash recover to CNI plugin

### PR DESCRIPTION
**Pull request by monsteredc**
This PR is to add the crash revocer like Calico for CNI plugin. The PR is to fix the issue #33137. 
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: cni/cmd/istio-cni/main.go [hunk 1]
```diff
@@ -118,7 +118,24 @@ func parseConfig(stdin []byte) (*PluginConf, error) {
...
+			msg := fmt.Sprintf("istio-cni panicked during ADD: %s", e)
+			if err != nil {
+				// If we're recovering and there was also an error, then we need to
+				// present both.
+				msg = fmt.Sprintf("%s: error=%s", msg, err)
+			}
+			err = fmt.Errorf(msg)
+		}
+		if err != nil {
+			log.Errorf("Final result of istio-cni ADD was an error.")
```
Comment (howardjohn) [inline cni/cmd/istio-cni/main.go:135]: can we log the error? and maybe the stack trace?

File: cni/cmd/istio-cni/main.go [hunk 2]
```diff
@@ -118,7 +118,24 @@ func parseConfig(stdin []byte) (*PluginConf, error) {
...
+func cmdAdd(args *skel.CmdArgs) (err error) {
+	// Defer a panic recover, so that in case if panic we can still return
+	// a proper error to the runtime.
+	defer func() {
+		if e := recover(); e != nil {
+			msg := fmt.Sprintf("istio-cni panicked during ADD: %s", e)
+			if err != nil {
+				// If we're recovering and there was also an error, then we need to
+				// present both.
+				msg = fmt.Sprintf("%s: error=%s", msg, err)
```
Comment (howardjohn) [inline cni/cmd/istio-cni/main.go:130]:
```suggestion
				msg = fmt.Sprintf("%s: %v", msg, err)
```

File: cni/cmd/istio-cni/main.go [hunk 3]
```diff
@@ -258,7 +275,24 @@ func cmdGet(args *skel.CmdArgs) error {
 }
 
 // cmdDel is called for DELETE requests
-func cmdDel(args *skel.CmdArgs) error {
+func cmdDel(args *skel.CmdArgs) (err error) {
+	// Defer a panic recover, so that in case we panic we can still return
```
[the remark this answers is not in the document]
Comment (monsteredc) [reply cni/cmd/istio-cni/main.go]: thanks!

File: cni/cmd/istio-cni/main.go (hunk 1 above)
Comment (monsteredc) [reply cni/cmd/istio-cni/main.go]: sounds great, thanks!

File: releasenotes/notes/33145.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: installation
```
Comment (bianpengyuan) [inline releasenotes/notes/33145.yaml:3]: I think it is fine to add this without a release notes.

File: cni/cmd/istio-cni/main.go [hunk 5]
```diff
@@ -118,7 +118,26 @@ func parseConfig(stdin []byte) (*PluginConf, error) {
...
 
 // cmdAdd is called for ADD requests
-func cmdAdd(args *skel.CmdArgs) error {
+func cmdAdd(args *skel.CmdArgs) (err error) {
+	// Defer a panic recover, so that in case if panic we can still return
+	// a proper error to the runtime.
+	defer func() {
+		if e := recover(); e != nil {
+			msg := fmt.Sprintf("istio-cni panicked during cmdAdd: %v", e)
+			log.Errorf(msg)
```
Comment (bianpengyuan) [inline cni/cmd/istio-cni/main.go:127]: I don't think we should log it here and at line 132. They were all duplicated. The recover block should only construct the crash message and the log should only happen at line 137.

File: releasenotes/notes/33145.yaml (hunk 4 above)
Comment (monsteredc) [reply releasenotes/notes/33145.yaml]: Alright, thanks!

File: cni/cmd/istio-cni/main.go (hunk 5 above)
Comment (monsteredc) [reply cni/cmd/istio-cni/main.go]: Thanks

**Review by bianpengyuan — APPROVED** [review]

**Pull request merged**